### PR TITLE
Drop secondary constructors

### PR DESCRIPTION
**Pull request by soc**
Missing bits:
- I need to implement moving statements in the class body to the
  constructor function to unbreak the Exception class and a few others.
- I need to understand why internal classes don't get constructors.
- ...

### CODE REVIEW

**Comment by soc** [comment]
This is work in progress for #43, and probably also #36.

**Comment by soc** [comment]
@dinfuehr 
```
$ RUST_BACKTRACE=1 target/debug/dora tests/array2.dora 
thread 'main' panicked at 'index out of bounds: the len is 0 but the index is 0', /rustc/bc44841ad2a2ad5f6c5e67b9e35ed8e7e71d4dc7/src/libcore/slice/mod.rs:2539:10
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   4: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:474
   5: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:381
   6: rust_begin_unwind
             at src/libstd/panicking.rs:308
   7: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
   8: core::panicking::panic_bounds_check
             at src/libcore/panicking.rs:61
   9: <usize as core::slice::SliceIndex<[T]>>::index
             at /rustc/bc44841ad2a2ad5f6c5e67b9e35ed8e7e71d4dc7/src/libcore/slice/mod.rs:2539
  10: core::slice::<impl core::ops::index::Index<I> for [T]>::index
             at /rustc/bc44841ad2a2ad5f6c5e67b9e35ed8e7e71d4dc7/src/libcore/slice/mod.rs:2396
  11: <alloc::vec::Vec<T> as core::ops::index::Index<I>>::index
             at /rustc/bc44841ad2a2ad5f6c5e67b9e35ed8e7e71d4dc7/src/liballoc/vec.rs:1677
  12: dora::baseline::expr::ExprGen::emit_assign
             at src/baseline/expr.rs:729
  13: dora::baseline::expr::ExprGen::emit_expr
             at src/baseline/expr.rs:143
  14: dora::baseline::expr::ExprGen::generate
             at src/baseline/expr.rs:126
  15: dora::baseline::codegen::CodeGen::emit_expr
             at src/baseline/codegen.rs:843
  16: dora::baseline::codegen::CodeGen::emit_stmt_expr
             at src/baseline/codegen.rs:651
  17: <dora::baseline::codegen::CodeGen<'a, 'ast> as dora_parser::ast::visit::Visitor<'ast>>::visit_stmt
             at src/baseline/codegen.rs:912
  18: dora::baseline::codegen::CodeGen::emit_stmt_block
             at src/baseline/codegen.rs:658
  19: <dora::baseline::codegen::CodeGen<'a, 'ast> as dora_parser::ast::visit::Visitor<'ast>>::visit_stmt
             at src/baseline/codegen.rs:920
  20: dora_parser::ast::visit::walk_fct
             at ./lib/dora-parser/src/ast/visit.rs:164
  21: dora_parser::ast::visit::Visitor::visit_fct
             at ./lib/dora-parser/src/ast/visit.rs:57
  22: dora::baseline::codegen::CodeGen::generate
             at src/baseline/codegen.rs:321
  23: dora::baseline::codegen::generate_fct
             at src/baseline/codegen.rs:87
  24: dora::baseline::codegen::generate
             at src/baseline/codegen.rs:49
  25: dora::baseline::dora_compile::patch_fct_call
             at src/baseline/dora_compile.rs:251
  26: dora::baseline::dora_compile::compile_request
             at src/baseline/dora_compile.rs:203
  27: <unknown>
fatal runtime error: failed to initiate panic, error 5
Aborted (core dumped)
```
This happens due to the field in `let field = &cls.fields[fieldid.idx()]` in `emit_assign` being empty. Any ideas?

**Comment by dinfuehr** [comment]
Yes, IIRC class and functions can't be generic both at the same time yet.

It could be that the Array constructor is a bit special, that length is initialized by the compiler. I will have to take a look into this in the evening...

**Comment by soc** [comment]
@dinfuehr Can you comment on why internal classes don't get constructors? I'm interested in understanding the requirements here.

**Comment by dinfuehr** [comment]
It seems that Array's constructors aren't internal methods. When allocating objects the memory area for an object is zero'ed and the classptr/vtableptr is written into the object's first word. For Array's also the length is written into the second word. This is necessary for fully initializing the object. After this the constructor is invoked with the object as its first argument.

**Comment by dinfuehr** [comment]
Object allocation is here: https://github.com/dinfuehr/dora/blob/master/src/baseline/asm.rs#L614

**Comment by soc** [comment]
Ah, thanks for the hints!

I now migrated the static functions to free-standing functions inside prelude for now, and I'm down to 18 test failures.

Regarding the handling of static functions and generics I have some ideas I'd like to discuss with you, but I'll open another issue for that as soon as this one is done.

**Comment by dinfuehr** [comment]
Before this PR gets too long or too much work, just split it into multiple PRs. E.g. one PR that just replaces constructors with static methods where this already works.

**Comment by dinfuehr** [comment]
Well, we could probably set length also in the Array-constructor. To me it feels to fit better to object allocation. Since the length is necessary to determine object size.

**Comment by soc** [comment]
I resolved all issues now, except the ones who rely on executing statements in the constructor: things involving Exception and Random.

> 312 tests passed; 4 tests ignored; 5 tests failed

**Comment by soc** [comment]
Basically we need to go through the fields and statements in the order the appear, and add their effects to the constructor:

```
class X(let msg: String) { // assign this first
    var i = 23;           // then this
    print(i);          // then this
}
```

At which place do you think this is done best?

**Comment by soc** [comment]
I implemented a very basic version of initializers, all tests are green now!

File: tests/char2.dora [hunk 1]
```diff
@@ -1,8 +1,8 @@
 fun main() {
-    let x = Array::<Char>();
-    assert(x.len() == 0);
+    //let x = arrayEmpty::<Char>();
```
Comment (dinfuehr) [inline tests/char2.dora]: I think this doesn't need to be commented out

File: src/class.rs [hunk 2]
```diff
@@ -53,7 +53,7 @@ pub struct Class {
     pub is_abstract: bool,
     pub internal: bool,
     pub internal_resolved: bool,
-    pub primary_ctor: bool,
+    pub has_constructor: bool,
 
     pub ctors: Vec<FctId>,
```
Comment (dinfuehr) [inline src/class.rs]: I think this should now also be `Option<FctId>` instead of `Vec<FctId>`

Comment (dinfuehr) [reply src/class.rs]: This should also make some code in typeck.rs unnecessary.

File: lib/dora-parser/src/parser.rs [hunk 3]
```diff
@@ -505,9 +462,8 @@ impl<'a> Parser<'a> {
                 }
 
                 _ => {
-                    return Err(MsgWithPos::new(self.lexer.path().to_string(),
-                                               self.token.position,
-                                               Msg::ExpectedClassElement(self.token.name())))
+                    let initializer = self.parse_expression_statement()?;
```
[the remark this answers is not in the document]
Comment (soc) [reply lib/dora-parser/src/parser.rs]: Yes, we could parse statements instead.

My idea for local variables in the constructor is the following:

```
class Foo(x: Int) {
  var field: Int = 0;
  {
    let x = 2;
    let y = 3;
    self.field = x * y;
  }
}
```

Based on my experience, this use-case is extremely rare though.

> Can you add tests for constructors with a bit more logic to test this?

Will do!

[the remark this answers is not in the document]
Comment (soc) [reply lib/dora-parser/src/parser.rs]: I added more tests and verified that the above syntax with local variables enclosed in braces works.

File: src/class.rs (hunk 2 above)
Comment (soc) [reply src/class.rs]: Fixed. Thanks!

File: tests/char2.dora (hunk 1 above)
Comment (soc) [reply tests/char2.dora]: Fixed!

**Comment by soc** [comment]
@dinfuehr Do you have any tasks left to do on this PR?

**Comment by dinfuehr** [comment]
Nope, didn't have time yet for a quick review.

Thanks a lot for the patch! I really like it a lot and a nice improvement for Dora :+1: I want to answer the issue about minor releases today as well.

**Comment by soc** [comment]
Thanks a lot!

Don't feel any pressure trying to answer my stuff before its convenient for you!

**Pull request merged**